packages/slate-react/src/plugin/react-editor.ts

```diff
@@ -412,6 +413,25 @@ export const ReactEditor: ReactEditorInterface = {
   },
 
   focus: editor => {
+    // If the editor has pending operations, focus should be requested after
+    // those changes are applied. Retry in the next tick if this is the case.
+    if (editor.operations.length > 0) {
+      const handle = FOCUS_TIMER_HANDLE.get(editor)
+      if (handle) {
+        clearTimeout(handle)
+        FOCUS_TIMER_HANDLE.delete(editor)
+      }
+      FOCUS_TIMER_HANDLE.set(
+        editor,
+        setTimeout(() => {
+          // Flush changes before trying to focus again
+          editor.onChange()
```

Comment:
Not sure if we should actually call onChange here, or just retry until there are no operations pending.

Reply:
onChange should already have been called by this point, right? I can't think of any situations where manually flushing the changes would be beneficial, and it might result in some weird edge cases. If, somehow, there are still pending operations after the timeout, retrying is probably safest.

Reply:
Do we still need a weak map if we're not calling onChange? I can't think of any adverse side effects of running the timeout callback twice if a consumer happens to call ReactEditor.focus twice.

Reply:
The only other thing that occurs to me is what happens if something goes wrong and editor.operations never gets emptied? That might happen if we call ReactEditor.focus after the editor has been unmounted, for example.

Perhaps we should add an `options: { retries?: number }` argument that defaults to 5 and is decremented on each attempt?

Reply:
@12joan - thank you for your comments. I've been looking a bit closer at this now and I'm unsure if this is actually the right fix, though it seems to help in some cases.

I think we might be able to just drop the whole selection concern in this function, as the selection change event handler should already have taken care of this. When we call `focus()` on the element, it should trigger a selection change event that will restore the proper selection.

I'll check in the new code which has this code removed, and that would definitely be the best if we don't actually need it here.

Reply:
> onChange should already have been called by this point, right? I can't think of any situations where manually flushing the changes would be beneficial, and it might result in some weird edge cases. If, somehow, there are still pending operations after the timeout, retrying is probably safest.

Yes, I have refactored this now. It will now call `onChange`, but only because it will try to set a selection in the top of the document if the editor currently isn't selected.

> Do we still need a weak map if we're not calling onChange? I can't think of any adverse side effects of running the timeout callback twice if a consumer happens to call ReactEditor.focus twice.

I agree. I have removed it - also one less WeakMap :)

> The only other thing that occurs to me is what happens if something goes wrong and editor.operations never gets emptied? That might happen if we call ReactEditor.focus after the editor has been unmounted, for example.
> Perhaps we should add an options: { retries?: number } argument that defaults to 5 and is decremented on each attempt?

Yeah, good thinking. I have implemented a retry (default 5) and then throw an error if the editor is still stuck with operations.